lgc/patch/Gfx9ConfigBuilder.cpp

```diff
@@ -1934,23 +1943,208 @@ template <typename T> void ConfigBuilder::buildPsRegConfig(ShaderStage shaderSta
+
+  assert(calcFactor.esVertsPerSubgroup == 1 && calcFactor.gsPrimsPerSubgroup == 1);
+  SET_REG_MOST_FIELD(&config->meshRegs, VGT_GS_ONCHIP_CNTL, ES_VERTS_PER_SUBGRP, 1);
+  SET_REG_MOST_FIELD(&config->meshRegs, VGT_GS_ONCHIP_CNTL, GS_PRIMS_PER_SUBGRP, 1);
+  SET_REG_MOST_FIELD(&config->meshRegs, VGT_GS_ONCHIP_CNTL, GS_INST_PRIMS_IN_SUBGRP, 1);
+  setNggSubgroupSize(1);
+
+  SET_REG_FIELD(&config->meshRegs, VGT_GS_PER_VS, GS_PER_VS, GsThreadsPerVsThread);
+
+  VGT_GS_OUTPRIM_TYPE gsOutputPrimitiveType = POINTLIST;
+  switch (meshMode.outputPrimitive) {
+  case OutputPrimitives::Points:
+    gsOutputPrimitiveType = POINTLIST;
+    break;
+  case OutputPrimitives::Lines:
+    gsOutputPrimitiveType = LINESTRIP;
+    break;
+  case OutputPrimitives::Triangles:
+    gsOutputPrimitiveType = TRISTRIP;
+    break;
+  default:
+    llvm_unreachable("Unknown primitive type!");
+    break;
+  }
+  SET_REG_FIELD(&config->meshRegs, VGT_GS_OUT_PRIM_TYPE, OUTPRIM_TYPE, gsOutputPrimitiveType);
+
+  const unsigned maxVertsPerSubgroup = std::min(meshMode.outputVertices, NggMaxThreadsPerSubgroup);
+  SET_REG_FIELD(&config->meshRegs, GE_MAX_OUTPUT_PER_SUBGROUP, MAX_VERTS_PER_SUBGROUP, maxVertsPerSubgroup);
+
+  //
+  // Build VS specific configuration
+  //
+  uint8_t usrClipPlaneMask = m_pipelineState->getRasterizerState().usrClipPlaneMask;
+  bool rasterizerDiscardEnable = m_pipelineState->getRasterizerState().rasterizerDiscardEnable;
+
+  SET_REG_FIELD(&config->meshRegs, PA_CL_CLIP_CNTL, UCP_ENA_0, (usrClipPlaneMask >> 0) & 0x1);
+  SET_REG_FIELD(&config->meshRegs, PA_CL_CLIP_CNTL, UCP_ENA_1, (usrClipPlaneMask >> 1) & 0x1);
+  SET_REG_FIELD(&config->meshRegs, PA_CL_CLIP_CNTL, UCP_ENA_2, (usrClipPlaneMask >> 2) & 0x1);
+  SET_REG_FIELD(&config->meshRegs, PA_CL_CLIP_CNTL, UCP_ENA_3, (usrClipPlaneMask >> 3) & 0x1);
+  SET_REG_FIELD(&config->meshRegs, PA_CL_CLIP_CNTL, UCP_ENA_4, (usrClipPlaneMask >> 4) & 0x1);
+  SET_REG_FIELD(&config->meshRegs, PA_CL_CLIP_CNTL, UCP_ENA_5, (usrClipPlaneMask >> 5) & 0x1);
+  SET_REG_FIELD(&config->meshRegs, PA_CL_CLIP_CNTL, DX_LINEAR_ATTR_CLIP_ENA, true);
+  SET_REG_FIELD(&config->meshRegs, PA_CL_CLIP_CNTL, DX_RASTERIZATION_KILL, rasterizerDiscardEnable);
+
+  SET_REG_FIELD(&config->meshRegs, PA_CL_VTE_CNTL, VPORT_X_SCALE_ENA, true);
+  SET_REG_FIELD(&config->meshRegs, PA_CL_VTE_CNTL, VPORT_X_OFFSET_ENA, true);
+  SET_REG_FIELD(&config->meshRegs, PA_CL_VTE_CNTL, VPORT_Y_SCALE_ENA, true);
+  SET_REG_FIELD(&config->meshRegs, PA_CL_VTE_CNTL, VPORT_Y_OFFSET_ENA, true);
+  SET_REG_FIELD(&config->meshRegs, PA_CL_VTE_CNTL, VPORT_Z_SCALE_ENA, true);
+  SET_REG_FIELD(&config->meshRegs, PA_CL_VTE_CNTL, VPORT_Z_OFFSET_ENA, true);
+  SET_REG_FIELD(&config->meshRegs, PA_CL_VTE_CNTL, VTX_W0_FMT, true);
+
+  SET_REG_FIELD(&config->meshRegs, PA_SU_VTX_CNTL, PIX_CENTER, 1);
+  SET_REG_FIELD(&config->meshRegs, PA_SU_VTX_CNTL, ROUND_MODE, 2); // Round to even
+  SET_REG_FIELD(&config->meshRegs, PA_SU_VTX_CNTL, QUANT_MODE, 5); // Use 8-bit fractions
```

Comment:
Can the PA_* register settings be shared with other geometry stages? (There are a few more below)

Reply:
Yes, I think so. I will factor out such piece of codes.

Reply:
> Can the PA_* register settings be shared with other geometry stages? (There are a few more below)

Fixed by https://github.com/GPUOpen-Drivers/llpc/pull/1951. Will rebase this PR after it.

Reply:
Thank you for doing this!